// dattatable/build/dashboard/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Navigation = void 0;
var gd_sprest_bs_1 = require("gd-sprest-bs");
var filter_1 = require("gd-sprest-bs/build/icons/svgs/filter");
/**
 * Navigation
 */
var Navigation = /** @class */ (function () {
    // Constructor
    function Navigation(props) {
        this._props = null;
        // Save the properties
        this._props = props;
        // Render the navigation
        this.render();
        // Call the render event
        props.onRendered ? props.onRendered(this._props.el) : null;
    }
    // Renders the component
    Navigation.prototype.render = function () {
        // Define the default props
        var props = {
            el: this._props.el,
            brand: this._props.title,
            enableSearch: this._props.hideSearch != null ? !this._props.hideSearch : null,
            items: this._props.items,
            itemsEnd: this._props.itemsEnd,
            searchBox: {
                hideButton: true,
                onChange: this._props.onSearch,
                onSearch: this._props.onSearch,
                placeholder: "Search this app",
            }
        };
        // Call the rendering event
        this._props.onRendering ? this._props.onRendering(props) : null;
        // Render a navbar
        var nav = gd_sprest_bs_1.Components.Navbar(props);
        // See if we are showing the filter
        if (this._props.hideFilter != true) {
            // Render the filter icon
            // Create a span to wrap the icon in
            var span = document.createElement("span");
            span.className = "bg-white d-inline-flex filter-icon ms-2 rounded";
            nav.el.firstElementChild.appendChild(span);
            // Render a tooltip
            var ttp = gd_sprest_bs_1.Components.Tooltip({
                el: span,
                content: "Filters",
                type: gd_sprest_bs_1.Components.TooltipTypes.Secondary,
                btnProps: {
                    // Render the filter button
                    iconType: filter_1.filter,
                    iconSize: 28,
                    type: gd_sprest_bs_1.Components.ButtonTypes.OutlineSecondary
                },
            });
            // Call the event
            this._props.onShowFilter ? ttp.el.addEventListener("click", this._props.onShowFilter) : null;
            this._props.onFilterRendered ? this._props.onFilterRendered(this._props.el) : null;
        }
        // Call the event
        this._props.onSearchRendered ? this._props.onSearchRendered(this._props.el.querySelector("input[type='search']")) : null;
    };
    return Navigation;
}());
exports.Navigation = Navigation;
